### PR TITLE
Fix and clean up GCSIO statistics logic

### PR DESCRIPTION
**Pull request by medb**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 1]
```diff
@@ -1595,8 +1588,7 @@ private String listStorageObjectsAndPrefixesPage(
 
     // Set delimiter if supplied.
     if (delimiter != null) {
-      listObject.setDelimiter(delimiter);
-      listObject.setIncludeTrailingDelimiter(includeTrailingDelimiter);
+      listObject.setDelimiter(delimiter).setIncludeTrailingDelimiter(true);
```
Comment (mprashanthsagar) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:1591]: Is this change(removing `includeTrailingDelimiter`) necessary ? 

Seems unrelated change, could we separate this into a different PR if it is intentional ?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          setHttpRequestFailureStats(httpResponse);
+        });
+  }
+
+  /** Set HTTP request stats */
+  private void setHttpRequestStats(HttpRequest request) {
+    switch (request.getRequestMethod()) {
+      case "DELETE":
+        increment(HTTP_DELETE_REQUEST);
+        break;
```
Comment (mprashanthsagar) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java:62]: Can we map `requestType` to `metric` and use a FunctionalInterface to increment instead ?

Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java:62]: I plan to change this logic to collect per-HTTP status statistics, so I will improve it then.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java (hunk 1 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java]: Reverted.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java [hunk 3]
```diff
@@ -439,6 +424,9 @@ GoogleCloudStorageItemInfo composeObjects(
       List<StorageResourceId> sources, StorageResourceId destination, CreateObjectOptions options)
       throws IOException;
 
+  /** Retrieves the statistics associated with the {@link GoogleCloudStorage} instance. */
+  Map<String, Long> getStatistics();
```
Comment (mprashanthsagar) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java:428]: nit : do we want to add a default impl ?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        increment(HTTP_PATCH_REQUEST);
+        break;
+      case "POST":
+        increment(HTTP_POST_REQUEST);
+        break;
+      case "PUT":
+        increment(HTTP_PUT_REQUEST);
+        break;
+      default:
+        throw new IllegalStateException("Unknown request method: " + request.getRequestMethod());
```
Comment (mprashanthsagar) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java:79]: Do we want to fail silently instead ?

**Review by mprashanthsagar — APPROVED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java (hunk 3 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java:428]: Ack, will follow up on this.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java (hunk 4 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StatisticsTrackingHttpRequestInitializer.java:79]: Ideally we would want to have stats for everything - hopefully our tests have sufficient coverage for all requests, but I probably will rework this anyway.
